In [1]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("Resources/charity_data.csv")
application_df

EIN                                               NAME  \
0       10520599                       BLUE KNIGHTS MOTORCYCLE CLUB   
1       10531628             AMERICAN CHESAPEAKE CLUB CHARITABLE TR   
2       10547893                 ST CLOUD PROFESSIONAL FIREFIGHTERS   
3       10553066                     SOUTHSIDE ATHLETIC ASSOCIATION   
4       10556103           GENETIC RESEARCH INSTITUTE OF THE DESERT   
...          ...                                                ...   
34294  996009318              THE LIONS CLUB OF HONOLULU KAMEHAMEHA   
34295  996010315           INTERNATIONAL ASSOCIATION OF LIONS CLUBS   
34296  996012607                                PTA HAWAII CONGRESS   
34297  996015768  AMERICAN FEDERATION OF GOVERNMENT EMPLOYEES LO...   
34298  996086871                           WATERHOUSE CHARITABLE TR   

      APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 12 columns]

## Model optimization

below you will see my three attempts to raise the model score from the previous note book. In this notebook you will see me copying the code from the Starter_Code notebook, and will be adjusting the input data to ensure that there are no variables or outliers that are causing confusion in the model.


In [2]:
#Drop un needed rows
application_df1 = application_df.drop(["EIN", "NAME"], axis=1)

In [3]:
application_df1.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

## First Attempt
Below you will see my first attempt to raiss the model score: I will change the binning threshold numbers to create more features in the models and perhaps increase the score.

In [4]:
application_df1["APPLICATION_TYPE"].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
application_df1["CLASSIFICATION"].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [6]:
'''
below if the original code that puts rare rows into categoriesin the APPLICATION_TYPE column. <= 1200 was the original number I will change it to 
500 resulting in more features after one-hot encoding
'''

type_df = application_df1.apply(lambda x: x.mask(x.map(x.value_counts())<=500, 'other') if x.name=='APPLICATION_TYPE' else x)
type_df


APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [7]:
'''
In the original code that puts rare rows into categories 
in the Classification column, was <= 4000 
was the original number. I will change it to 
7000 resulting in more features after one-hot encoding
'''

application_df2 = type_df.apply(lambda x: x.mask(x.map(x.value_counts())<=4000, 'other') if x.name=='CLASSIFICATION' else x)




In the previous notebook the one-hot encoding reulted in 36 features. below there is an increase in columns to 42.

In [8]:
numeric_df = pd.get_dummies(application_df2)
numeric_df

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  \
0           1      5000              1                     1   
1           1    108590              1                     0   
2           1      5000              0                     0   
3           1      6692              1                     0   
4           1    142590              1                     0   
...       ...       ...            ...                   ...   
34294       1      5000              0                     0   
34295       1      5000              0                     0   
34296       1      5000              0                     0   
34297       1      5000              1                     0   
34298       1  36500179              0                     0   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                         0                    0                    0   
1                         0                    1                    0   
2                         0                    0                    0   
3                         0                    1                    0   
4                         0                    1                    0   
...                     ...                  ...                  ...   
34294                     0                    0                    1   
34295                     0                    0                    1   
34296                     0                    1                    0   
34297                     0                    0                    0   
34298                     0                    1                    0   

       APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  ...  \
0                        0                    0                    0  ...   
1                        0                    0                    0  ...   
2                        1                    0                    0  ...   
3                        0                    0                    0  ...   
4                        0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    0                    0                    0  ...   
34295                    0                    0                    0  ...   
34296                    0                    0                    0  ...   
34297                    1                    0                    0  ...   
34298                    0                    0                    0  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                      0                       0                         0   
1                      1                       0                         0   
2                      0                       0                         0   
3                      0                       1                         0   
4                      0                       0                         1   
...                  ...                     ...                       ...   
34294                  0                       0                         0   
34295                  0                       0                         0   
34296                  0                       0                         0   
34297                  0                       0                         0   
34298                  0                       0                         0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                       0                 0                       0   
1                       0                 0                       0   
2                       0                 0                       0   
3                       0                 0                       0   
4                       0                 0                       0   
...                   ...               ...                     ...   
34

In [9]:
# Split our preprocessed data into our features and target arrays
#  YOUR CODE GOES HERE
y = numeric_df["IS_SUCCESSFUL"]
X = numeric_df.drop("IS_SUCCESSFUL", axis=1)



# Split the preprocessed data into a training and testing dataset
#  YOUR CODE GOES HERE

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [10]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [11]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_model = tf.keras.models.Sequential()

# First hidden layer
#  YOUR CODE GOES HERE

nn_model.add(tf.keras.layers.Dense(units=15, activation="relu", input_dim=41))

# Second hidden layer
#  YOUR CODE GOES HERE
nn_model.add(tf.keras.layers.Dense(units=25, activation="relu"))

# Output layer
#  YOUR CODE GOES HERE
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 15)                630       
                                                                 
 dense_1 (Dense)             (None, 25)                400       
                                                                 
 dense_2 (Dense)             (None, 1)                 26        
                                                                 
Total params: 1,056
Trainable params: 1,056
Non-trainable params: 0
_________________________________________________________________


2022-08-06 19:54:20.105259: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [12]:
# Compile the model
#  YOUR CODE GOES HERE

nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [13]:
# Train the model
#  YOUR CODE GOES HERE
fit_model = nn_model.fit(X_train_scaled,y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5874 - accuracy: 0.7118
Epoch 2/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5615 - accuracy: 0.7247
Epoch 3/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5583 - accuracy: 0.7266
Epoch 4/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5567 - accuracy: 0.7274
Epoch 5/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5557 - accuracy: 0.7283
Epoch 6/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5545 - accuracy: 0.7288
Epoch 7/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5536 - accuracy: 0.7291
Epoch 8/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5531 - accuracy: 0.7292
Epoch 9/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5526 - accuracy: 0.7301
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5522 - accura

804/804 [==============================] - 1s 2ms/step - loss: 0.5442 - accuracy: 0.7352
Epoch 82/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5441 - accuracy: 0.7348
Epoch 83/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5440 - accuracy: 0.7352
Epoch 84/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5445 - accuracy: 0.7327
Epoch 85/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5437 - accuracy: 0.7348
Epoch 86/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5439 - accuracy: 0.7351
Epoch 87/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5441 - accuracy: 0.7358
Epoch 88/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5442 - accuracy: 0.7345
Epoch 89/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5437 - accuracy: 0.7346
Epoch 90/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5438 - accuracy: 

In [14]:
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5627 - accuracy: 0.7258 - 377ms/epoch - 1ms/step
Loss: 0.5626850724220276, Accuracy: 0.7258309125900269


## Attempt 1 complete
changing the binning numbers made very little difference to the final score. original score was 0.7219 and current is 0.7258 the difference in minimal.

## Second attempt

Below you will see my second attempt at improving the tensorflow model. I changed activation function from relu to tanh, I added a another hidden layer, I changed node numbers to 10,20,30 and changed epoch to 50.

In [31]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_model = tf.keras.models.Sequential()

# First hidden layer
#  YOUR CODE GOES HERE

nn_model.add(tf.keras.layers.Dense(units=10, activation="tanh", input_dim=41))

# Second hidden layer
#  YOUR CODE GOES HERE
nn_model.add(tf.keras.layers.Dense(units=20, activation="tanh"))
nn_model.add(tf.keras.layers.Dense(units=30, activation="tanh"))
# Output layer
#  YOUR CODE GOES HERE
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_14 (Dense)            (None, 10)                420       
                                                                 
 dense_15 (Dense)            (None, 20)                220       
                                                                 
 dense_16 (Dense)            (None, 30)                630       
                                                                 
 dense_17 (Dense)            (None, 1)                 31        
                                                                 
Total params: 1,301
Trainable params: 1,301
Non-trainable params: 0
_________________________________________________________________


In [32]:
# Compile the model
#  YOUR CODE GOES HERE

nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [33]:
# Train the model
#  YOUR CODE GOES HERE
fit_model = nn_model.fit(X_train_scaled,y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5864 - accuracy: 0.7121
Epoch 2/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5650 - accuracy: 0.7251
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5613 - accuracy: 0.7264
Epoch 4/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5593 - accuracy: 0.7255
Epoch 5/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5584 - accuracy: 0.7255
Epoch 6/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5569 - accuracy: 0.7263
Epoch 7/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5565 - accuracy: 0.7269
Epoch 8/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5550 - accuracy: 0.7292
Epoch 9/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5548 - accuracy: 0.7280
Epoch 10/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5545 - accuracy: 0.7285

In [34]:
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5595 - accuracy: 0.7261 - 364ms/epoch - 1ms/step
Loss: 0.5594861507415771, Accuracy: 0.726064145565033


## Result 